### ISSUE TITLE
Off-by-one error in reported epoch when restarting training

### ISSUE THREAD

**Issue by oliverwm1** [ai2cm/ace#183]
Jobs which are stopped and restarted result in an off-by-one error in the reported epoch. Otherwise training proceeds fine.

For example, here is the epoch reported to wandb for two runs. The gray one was restarted after epoch=1. "epoch=1" is reported twice resulting in the reported epoch being too small by 1 after restarting.

![Screen Shot 2023-08-14 at 4 17 50 PM](https://github.com/ai2cm/ace/assets/23640870/5efd27ca-703f-4495-bc90-5f2f816b170a)

Possible solution: instead of manually serializing the current epoch and num_batches_seen (see below) we can defer to the version of these quantities which are already serialized by the optimizer. This would ensure there is a single source of truth and I believe would resolve the issue except for the much rarer case of job getting interrupted between the scheduler stepping and the checkpoint saving.

https://github.com/ai2cm/full-model/blob/0e17e1b17edaddf48bbecd78747779972ff6ae79/fme/fme/fcn_training/train.py#L302-L309

**Comment by mcgibbon** [comment]
> except for the much rarer case of job getting interrupted between the scheduler stepping and the checkpoint saving

This wouldn't actually be an edge case, if that happened we'd just end up loading the checkpoint from before the scheduler stepped. We just need to ensure all state is consistent at the time when the checkpoint is saved. It wasn't before, since we incremented epoch after saving at the start of the next loop instead of before at the end of the current loop.

**Comment by mcgibbon** [comment]
Ohhh interesting, you mean the wandb report would still show this. I don't think it would, though it will report the same data (if we're reproducible) twice for the same epoch. But I'm curious what wandb will do in that case. I hope "overwrite" the data instead of crashing or something.

**Comment by mcgibbon** [comment]
This also makes me realize we don't have "reproducible restarts" since the current random seed isn't saved/restored when resuming, but it probably isn't important.

**Comment by oliverwm1** [comment]
This is still happening. I have three jobs that restarted twice partway through, and they all had the following behavior:

![W B Chart 9_14_2023, 2_28_31 PM](https://github.com/ai2cm/ace/assets/23640870/fbf7428a-2dd8-4385-8b4f-d135ed164730)

([wandb run](https://wandb.ai/ai2cm/fourcastnet/runs/hsp6v3n6?workspace=user-oliverwm))

Note these runs used 93b8e518baf623dd1d7d75972e0001e686f5a104 which was after fix was merged.

**Comment by mcgibbon** [comment]
I think the issue is `self.epoch` needs to be incremented before `save_epoch` is called in the training loop, not just at the end vs start of the training loop.

A safer way to handle this might be to separate this idea of "loop epoch" from "model epoch", and increment the model epoch when train_one_batch is called. `self.epoch` could be the model epoch. We still need a separate "loop epoch" variable so e.g. the time for the epoch is logged with the right epoch number.

**Comment by mcgibbon** [comment]
OK, should be fixed for real this time https://github.com/ai2cm/full-model/pull/353
